### PR TITLE
Enable caches for linux_x64

### PR DESCRIPTION
**Pull request by sbogolepov**

### CODE REVIEW

**Review by homuroll — APPROVED** [review]

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt [hunk 1]
```diff
@@ -112,14 +112,19 @@ internal val IrClass.kotlinObjCClassInfoSymbolName: String
...
+fun IrFunction.computeSymbolName() = with(KonanBinaryInterface) { symbolName }.replaceSpecialSymbols()
 
-val IrClass.typeInfoSymbolName get() = with(KonanBinaryInterface) { typeInfoSymbolName }
+fun IrField.computeSymbolName() = with(KonanBinaryInterface) { symbolName }.replaceSpecialSymbols()
+
+fun IrClass.computeTypeInfoSymbolName() = with(KonanBinaryInterface) { typeInfoSymbolName }.replaceSpecialSymbols()
+
+private fun String.replaceSpecialSymbols() =
+        // '@' is used for symbol versioning in GCC: https://gcc.gnu.org/wiki/SymbolVersioning.
+        this.replace("@", "__")
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt:127]: So here we generally hope that this won't make symbol names clash?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/DataFlowIR.kt [hunk 2]
```diff
@@ -639,7 +638,7 @@ internal object DataFlowIR {
 
             assert(irField.parent !is IrClass) { "All local properties initializers should've been lowered" }
             val attributes = FunctionAttributes.IS_GLOBAL_INITIALIZER or FunctionAttributes.RETURNS_UNIT
-            val symbol = FunctionSymbol.Private(privateFunIndex++, module, -1, attributes, null, null, takeName { "${irField.symbolName}_init" })
+            val symbol = FunctionSymbol.Private(privateFunIndex++, module, -1, attributes, null, null, takeName { "${irField.computeSymbolName()}_init" })
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/DataFlowIR.kt:641]: (This change probably belongs to other commit).

File: konan/konan.properties [hunk 3]
```diff
@@ -400,14 +401,16 @@ targetSysRoot.linux_x64 = $gccToolchain.linux_x64/x86_64-unknown-linux-gnu/sysro
...
 targetCpu.linux_x64 = x86-64
-clangFlags.linux_x64 = -cc1 -target-cpu $targetCpu.linux_x64 -emit-obj -disable-llvm-optzns -x ir
+clangFlags.linux_x64 = -cc1 -target-cpu $targetCpu.linux_x64 -emit-obj -disable-llvm-optzns -x ir \
+   -ffunction-sections -fdata-sections
 clangNooptFlags.linux_x64 = -O1
-clangOptFlags.linux_x64 = -O3 -ffunction-sections
+clangOptFlags.linux_x64 = -O3
 clangDebugFlags.linux_x64 = -O0
 clangDynamicFlags.linux_x64 = -mrelocation-model pic
+clangStaticFlags.linux_x64 = -mrelocation-model pic
```
Comment (SvyatoslavScherbina) [inline konan/konan.properties:410]: Might be related: `OptimizationPipeline.kt` also has relocation mode. It is set to "default", and AFAIK it is not PIC on Linux.

File: konan/konan.properties [hunk 4]
```diff
@@ -68,7 +68,8 @@ cacheableTargets.macos_x64 = \
   ios_x64 \
   ios_arm64
 
-cacheableTargets.linux_x64 =
+cacheableTargets.linux_x64 = \
+  linux_x64
```
Comment (SvyatoslavScherbina) [inline konan/konan.properties:72]: Btw, did hidden visibility (`makeVisibilityHiddenLikeLlvmInternalizePass`) for static caches work as expected? I mean, does it help to make cached symbols non-exported in final executable or .so?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt (hunk 1 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt:127]: Yes. The tricky part here are functions that contain `@` in their name. Maybe make it more robust and readable, e.g. `__at__`?

File: konan/konan.properties (hunk 3 above)
Comment (sbogolepov) [reply konan/konan.properties:410]: Yes, good catch. Looks like I'll have to add a separate properties for relocation mode.

File: konan/konan.properties (hunk 4 above)
Comment (sbogolepov) [reply konan/konan.properties:72]: Yes. More than that, it helps linker to cope with absence of some runtime declarations like `Kotlin_ObjCInterop_getTypeInfoForClass`.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt (hunk 1 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/BinaryInterface.kt:127]: Well, maybe. On the other hand, functions that contain `@` in their name probably could clash with extension functions before this patch. So let's avoid complicated mangling for now. `__at__` or similar things should be just ok.

File: konan/konan.properties (hunk 3 above)
Comment (sbogolepov) [reply konan/konan.properties]: Done.

**Review by SvyatoslavScherbina — APPROVED** [review]
Everything [else](https://github.com/JetBrains/kotlin-native/pull/4596#discussion_r544097648) is ok.

**Pull request merged**